src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,290 @@
+                // If we didn't have it in our cache, see if we can load it from disk.
+                // Note: pass 'loadOnly' so we only attempt to load from disk, not to actually
+                // try to create the metadata.
+                var info = await SymbolTreeInfo.TryGetInfoForMetadataAssemblyAsync(
+                    solution, assembly, reference, loadOnly: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+                return info;
+            }
 
-                var info = SymbolTreeInfo.Create(VersionStamp.Default, compilation.Assembly, cancellationToken);
-                if (info != null)
+            public async Task<SymbolTreeInfo> TryGetSymbolTreeInfoAsync(
+                Project project, CancellationToken cancellationToken)
+            {
+                ProjectInfo projectInfo;
+                if (_projectToInfo.TryGetValue(project.Id, out projectInfo))
                 {
-                    RecordCount(project.Id, info.Count);
+                    var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+                    if (version == projectInfo.VersionStamp)
+                    {
+                        return projectInfo.SymbolTreeInfo;
+                    }
                 }
+
+                return null;
+            }
+        }
+
+        private class IncrementalAnalyzer : IncrementalAnalyzerBase
+        {
+            private readonly ConcurrentDictionary<ProjectId, ProjectInfo> _projectToInfo;
+
+            // Note: the Incremental-Analyzer infrastructure guarantees that it will call all the methods
+            // on this type in a serial fashion.  As such, we don't need explicit locking, or threadsafe
+            // collections (if they're only used by this type).  So, for example, the map we populate
+            // needs to be a ConcurrentDictionary as it will be read and written from multiple types.
+            // However, the HashSet<ProjectId> is ok as it will only be used by this type and there is
+            // no concurrency in this type on its own.
+            private readonly ConcurrentDictionary<string, MetadataInfo> _metadataPathToInfo;
+
+            private readonly Dictionary<ProjectId, Project> idTolastSeenProject = new Dictionary<ProjectId, Project>();
+
+            public IncrementalAnalyzer(
+                ConcurrentDictionary<ProjectId, ProjectInfo> projectToInfo,
+                ConcurrentDictionary<string, MetadataInfo> metadataPathToInfo)
+            {
+                _projectToInfo = projectToInfo;
+                _metadataPathToInfo = metadataPathToInfo;
             }
 
-            public override Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken)
+            public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
```

Comment:
@heejaechang When does 'AnalyzeProjectAsync' get called?  Is it interleaved with document analyzing?  Or does it all happen at the end?